### ISSUE TITLE
Cannot use rumble with controllers on windows.

### ISSUE THREAD

**Issue by DragonMoffon** [pyglet/pyglet#574]
The only way to have a controller rumble is through pyglet.input.evdev.py. In the standard definition of GameContoller, the rumble_play_weak() and rumble_play_strong() methods are left empty. They are only modified to use rumble in the EvdevGameController child class of GameController. However, this definition relies on fcntl module, which (to the best of my knowledge) is not supported on windows, which means that there is no way to use pyglet to have a controller rumble on windows.

**Comment by benmoran56** [comment]
Hi, thanks for opening the ticket. The Controller module is going to be new in 2.0, and so far only Linux has rumble implemented. As you've experienced, the `evdev` module is Linux specific. 

Currently on Windows, the Controller/Joystick APIs are using the DirectInput backend. Support for rumble can be added, but bindings will need to be written.
We will also likely add a native Xinput backed for those Controllers that support it, which will include rumble.

Nobody is currently working on it, but I will modify this ticket and mark it as an enhancement.

**Comment by benmoran56** [comment]
@DragonMoffon 
Just wanted to give you a heads-up that we added XInput support for Controllers on Windows. 
This includes Rumble support. 
Please give it a try, and let us know how it works out for you. 
If you haven't seen it, there is a Controller example in the repo, at `examples/input/controller.py` which should play rumble effects when pressing buttons.

**Comment by DragonMoffon** [comment]
I've found a few weird issues with the example.py. It may be my fault, as I did pull this controller apart to clean it and may have somehow crossed some wires. It could also be that because I am using a Dualshock 4 that is being passed through the steam system.

Firstly the rumble didn't work.
Secondly, the axis of the joysticks was swapped (In the display).
Thirdly, this is where the wires may have been crossed, one of my triggers is read as the 'x'-axis of one of the joysticks, and the 'x'-axis of the joystick acts like the trigger input.


https://user-images.githubusercontent.com/86714785/163796264-242358d2-bc42-4750-b6ce-f704e36d13c3.mp4

**Comment by caffeinepills** [comment]
Just to add, we added XInput support, which is for Xbox-type controllers. There are programs that convert a controller to an XInput device (like ds4 to xinput wrappers).

1) It won't work unless it's an xinput device.
2) Not sure about this one
3) I would check with the OS game controller preview to make sure they appear correctly.

**Comment by DragonMoffon** [comment]
Yeah, it detects a ps4 controller. Which is probably why it doesn't work. I currently don't have an xbox controller so I can't test that, unfortunately.

**Comment by einarf** [comment]
Works flawlessly for me with xbox controller

**Comment by benmoran56** [comment]
@DragonMoffon your PS4 controller can still work with the DirectInput backend, you just won't have rumble. 
It seems like your mappings are off. Maybe this is because they were remapped in Steam?
Can you check the following code? 
```py
import pyglet

controller = pyglet.input.get_controllers()[0]
print(controller.name, controller.guid)

mapping = pyglet.input.controller.get_mapping(controller.guid)
print(mapping)
```

**Comment by DragonMoffon** [comment]
@benmoran56 

```python
PS4 Controller 030000004c050000cc09000000000000
{
 'guid': '030000004c050000cc09000000000000', 
 'name': 'PS4 Controller', 
 'a': Relation(type=button, index=1, inverted=False),
 'b': Relation(type=button, index=2, inverted=False),
 'back': Relation(type=button, index=8, inverted=False), 
 'dpdown': Relation(type=hat0, index=4, inverted=False),
 'dpleft': Relation(type=hat0, index=8, inverted=False),
 'dpright': Relation(type=hat0, index=2, inverted=False),
 'dpup': Relation(type=hat0, index=1, inverted=False),
 'guide': Relation(type=button, index=12, inverted=False),
 'leftshoulder': Relation(type=button, index=4, inverted=False),
 'leftstick': Relation(type=button, index=10, inverted=False), 
 'lefttrigger': Relation(type=axis, index=3, inverted=False),
 'leftx': Relation(type=axis, index=0, inverted=False),
 'lefty': Relation(type=axis, index=1, inverted=False),
 'rightshoulder': Relation(type=button, index=5, inverted=False),
 'rightstick': Relation(type=button, index=11, inverted=False),
 'righttrigger': Relation(type=axis, index=4, inverted=False), 
 'rightx': Relation(type=axis, index=2, inverted=False),
 'righty': Relation(type=axis, index=5, inverted=False),
 'start': Relation(type=button, index=9, inverted=False),
 'x': Relation(type=button, index=0, inverted=False),
 'y': Relation(type=button, index=3, inverted=False)
}
```

**Comment by benmoran56** [comment]
Looks like you have one of the "v2" DualShock 4s. You can download newer controller mappings from here, and add them to your game on startup: https://github.com/gabomdq/SDL_GameControllerDB
```py
import pyglet

pyglet.input.controller.add_mappings_from_file("gamecontrollerdb.txt")
```

**Comment by benmoran56** [comment]
@DragonMoffon Did you have a chance to try after updating the mapping DB?

**Comment by DragonMoffon** [comment]
@benmoran56 I just did a little test, and I can't seem to get pyglet to recognise the controller at all. It seems to be a problem with steam as it works perfectly with ds4 windows. When using ds4 windows I get full rumble control and everything so that works great! I don't know how to use the custom bindings. I used the function call you provided, but it doesn't seem to do anything.

**Comment by benmoran56** [comment]
Sorry, could you clarify what you mean? I'm not sure I understand. 
Works OK on Windows now, but still not working on OSX? 

> I can't seem to get pyglet to recognise the controller at all.

Do you mean on OSX? Can you confirm that the standard Joystick interface works?

**Comment by DragonMoffon** [comment]
All tests have been done on Windows 11.
I have a ps4 controller plugged into my computer and when using steam pyglet does not recognise the controller at all. When using DS4windows the controller works perfectly. I have not tested the joystick interface. I have no OSX machine I could test the system out on.

**Comment by benmoran56** [comment]
Ah, now I understand. Thank you.

Can you see what the following code produces, under both circumstances?

```py
for js in pyglet.input.get_joysticks():
    print(js.device.name, js.device.get_guid())
```
Controllers are a subset of Joysticks, so this should hopefully give us more information.

**Comment by caffeinepills** [comment]
This is now supported in 3.0 via Mac's GameController API. It does have limited controller support as they mostly only offer it to more popular commercial controllers like Xbox, Playstation and MFi gamepads. Never the less, it's the best Apple will give us.
